service/history/callback_queue_factory.go

```diff
@@ -0,0 +1,206 @@
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package history
+
+import (
+	"net/http"
+	"sync"
+
+	"go.uber.org/fx"
+
+	"go.temporal.io/server/common/log"
+	"go.temporal.io/server/common/log/tag"
+	"go.temporal.io/server/common/metrics"
+	"go.temporal.io/server/common/quotas"
+	ctasks "go.temporal.io/server/common/tasks"
+	"go.temporal.io/server/service/history/queues"
+	"go.temporal.io/server/service/history/shard"
+	"go.temporal.io/server/service/history/tasks"
+	wcache "go.temporal.io/server/service/history/workflow/cache"
+)
+
+const callbackQueuePersistenceMaxRPSRatio = 0.3
+
+type callbackQueueFactoryParams struct {
+	fx.In
+
+	QueueFactoryBaseParams
+}
+
+type completionClientPool struct {
+	mu      sync.RWMutex
+	callers map[queues.NamespaceIDAndDestination]HTTPCaller
+}
+
+func (p *completionClientPool) getCaller(namespaceID, destination string) HTTPCaller {
+	key := queues.NamespaceIDAndDestination{NamespaceID: namespaceID, Destination: destination}
+	p.mu.RLock()
+	caller, ok := p.callers[key]
+	p.mu.RUnlock()
+	if !ok {
+		p.mu.Lock()
+		if caller, ok = p.callers[key]; !ok {
+			// In the future, we'll want to support HTTP2 clients as well.
+			client := &http.Client{}
+			caller = client.Do
+		}
+		p.mu.Unlock()
+	}
+
+	return caller
+}
+
+type callbackQueueFactory struct {
+	callbackQueueFactoryParams
+	hostReaderRateLimiter quotas.RequestRateLimiter
+	// Shared client pool for all shards in the host.
+	completionCallerPool completionClientPool
+	// Shared scheduler across all shards in the host.
+	hostScheduler queues.Scheduler
+}
+
+func NewCallbackQueueFactory(params callbackQueueFactoryParams) QueueFactory {
+	grouper := queues.GrouperNamespaceIDAndDestination{}
+	return &callbackQueueFactory{
+		callbackQueueFactoryParams: params,
+		hostReaderRateLimiter: queues.NewReaderPriorityRateLimiter(
+			NewHostRateLimiterRateFn(
+				params.Config.CallbackProcessorMaxPollHostRPS,
+				params.Config.PersistenceMaxQPS,
+				callbackQueuePersistenceMaxRPSRatio,
+			),
+			int64(params.Config.TransferQueueMaxReaderCount()),
+		),
+		completionCallerPool: completionClientPool{
+			callers: make(map[queues.NamespaceIDAndDestination]HTTPCaller, 0),
+		},
+		hostScheduler: &queues.CommonSchedulerWrapper{
+			Scheduler: ctasks.NewMultiDestinationScheduler[queues.NamespaceIDAndDestination, queues.Executable](
+				ctasks.MultiDestinationSchedulerOptions[queues.NamespaceIDAndDestination, queues.Executable]{
+					KeyFn: func(e queues.Executable) queues.NamespaceIDAndDestination { return grouper.KeyTyped(e.GetTask()) },
+					LimiterOptionsFn: func(key queues.NamespaceIDAndDestination) ctasks.LimiterOptions {
+						// TODO: get these from dynamic config.
+						return ctasks.LimiterOptions{
+							Concurrency: func() int { return 10 },
+							RateLimiter: quotas.NewDefaultOutgoingRateLimiter(func() float64 { return 100.0 }),
```

Comment:
There's [an existing rate limiter](https://github.com/temporalio/temporal/blob/main/service/history/queue_factory_base.go#L155) for controlling the task processing rate across all task categories.  

I think we should make the RateLimiter here to work with the existing overall rate limiter, probably via a [MultiRequestRateLimiter](https://github.com/temporalio/temporal/blob/main/common/quotas/multi_request_rate_limiter_impl.go).

Reply:
I'm not sure we want to mix these up TBH. What's the justification for it?